archival: Add offset_range_size method to the storage::log

Lazin opened the pull request:
This PR adds a `offset_range_size` methods to the `storage::log`. The implementation in `storage::disk_log_impl` uses segment indexes to find approximate location of the start of the range and approximate location of the last batch. Then it performs short scan to find the precise location.

There are two overloads of the method.
- First one accepts two offsets and returns the size of the range.
- Second one accepts start offset and target size. It finds the last offset of the range and its size.

The method is supposed to be used to create S3 uploads. Currently, the code reads from disk directly and therefore is coupled with the storage format and a lot of internal storage details. Our goal is to decouple local and remote storage by providing clear API boundary.


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

VladLazar commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,537 @@
+            _parent->_buffer.append(std::move(rec_iobuf));
+            _parent->_num_batches++;
+            bool max_bytes_reached = _parent->_buffer.size_bytes()
+                                     > (size_t)_parent->_max_bytes;
```
nit: `static_cast`

VladLazar commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,537 @@
+            ss::temporary_buffer<char> empty;
+            co_return empty;
+        }
+        auto top = _buffer.begin();
+        auto buf = std::move(*top).release();
```
nit: `head` and `head_buf` might be more fitting names

VladLazar commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,537 @@
+
+        // Left side
+        subscan_res left_scan;
+        {
```
Could we factor these partial scans into a different function?

VladLazar reviewed (commented):
Looks neat. I understand avoiding the batch cache, but why not go through the readers cache?

nvartolomei commented on src/v/archival/async_data_uploader.h:
```diff
@@ -0,0 +1,178 @@
+    }
+
+private:
+    void throw_if_not_initialized(std::string_view caller) const;
```
I assume you're splitting the constructor and initialization because you need initialization to be async? Why not make the constructor private and have a static async function that constructs the object and initializes it properly? This way you wouldn't have to add guards everywhere and will be impossible to have a "non-initialized" `segment_upload` object.

nvartolomei commented on src/v/archival/async_data_uploader.h:
```diff
@@ -0,0 +1,178 @@
+} // namespace detail
+
+/// Representation of the inclusive monotonic offset range
+struct inclusive_offset_range {
```
I'd prefer having an `offset_range` type, and if you want to avoid having to do the `x + 1` math everywhere, add a static method that builds an inclusive range so you can call it like `offset_range::inclusive(a, b)`.

One reason is consistency with all kind of ranges in C++.
Second reason is that `size()` will begin to make sense for default constructed ranges. Currently, `inclusive_offset_range{}.size() == 1` which doesn't make sense.
Third one,

nvartolomei commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,537 @@
+  model::timeout_clock::duration timeout,
+  ss::gate::holder h,
+  retry_chain_logger& rtc) {
+    auto deadline = model::timeout_clock::now() + timeout;
```
I'm curious why we are mixing timeouts (which are propagated down to very fine grained operations) with deadlines?

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/38576#018b15bf-a837-4a2d-bb9e-9dcc405bc341

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40235#018b8b12-6e0e-4316-b2d3-ad6cd8e762ba: "rptest.tests.cluster_features_test.FeaturesSingleNodeTest.test_get_features"
"rptest.tests.read_replica_e2e_test.ReadReplicasUpgradeTest.test_upgrades.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.upgrade_test.UpgradeFromPriorFeatureVersionTest.test_basic_upgrade"
"rptest.tests.upgrade_test.RedpandaInstallerTest.test_install_by_line"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_move_consumer_offsets_intranode.num_to_upgrade=2"
"rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade_with_rollback.upgrade_after_rollback=True"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_cancel_ongoing_movements"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_rebalancing_node.shutdown_decommissioned=False"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=True.tick_interval=3600000"
"rptest.tests.nodes_decommissioning_test.NodeDecommissionFailureReportingTest.test_allocation_failure_reporting"
"rptest.tests.acls_test.AccessControlListTestUpgrade.test_upgrade_sasl"
"rptest.tests.cluster_bootstrap_test.ClusterBootstrapUpgrade.test_change_bootstrap_configs_during_upgrade.empty_seed_starts_cluster=False"
"rptest.tests.cluster_config_test.ClusterConfigLegacyDefaultTest.test_legacy_default_explicit_after_upgrade.wipe_cache=False"
"rptest.tests.cluster_features_test.FeaturesMultiNodeTest.test_explicit_activation"
"rptest.tests.cluster_features_test.FeaturesMultiNodeUpgradeTest.test_upgrade"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_empty.num_to_upgrade=2"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40235#018b8b12-6e11-49da-acf9-e7d2d52e0588: "rptest.tests.cluster_features_test.FeaturesSingleNodeUpgradeTest.test_upgrade"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_bootstrapping_after_move.num_to_upgrade=2"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_overlapping_changes.num_to_upgrade=2"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart.new_bootstrap=False"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_crashed_node"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_rebalancing_node.shutdown_decommissioned=True"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=True.tick_interval=5000"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recommissioning_node_finishes"
"rptest.tests.cluster_features_test.FeaturesNodeJoinTest.test_old_node_join"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_static.num_to_upgrade=2"
"rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False"
"rptest.tests.cluster_bootstrap_test.ClusterBootstrapUpgrade.test_change_bootstrap_configs_during_upgrade.empty_seed_starts_cluster=True"
"rptest.tests.cluster_features_test.FeaturesMultiNodeTest.test_get_features"
"rptest.tests.cluster_config_test.ClusterConfigLegacyDefaultTest.test_legacy_default_explicit_after_upgrade.wipe_cache=True"
"rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade"
"rptest.tests.license_upgrade_test.UpgradeToLicenseChecks.test_basic_upgrade.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.pandaproxy_test.BasicAuthUpgradeTest.test_upgrade_and_enable_basic_auth.base_release=.22.2.next_release=.22.3"
"rptest.tests.offset_retention_test.OffsetRetentionDisabledAfterUpgrade.test_upgrade_from_pre_v23.initial_version=.22.2.9"
"rptest.tests.upgrade_test.UpgradeFromSpecificVersion.test_basic_upgrade"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40235#018b8b12-6e15-461b-98f6-f3f4add83f13: "rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=3600000"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_finishes_after_manual_cancellation.delete_topic=False"
"rptest.tests.partition_movement_upgrade_test.PartitionMovementUpgradeTest.test_basic_upgrade"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_flipping_decommission_recommission.node_is_alive=False"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recommissioning_one_of_decommissioned_nodes"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_dynamic.num_to_upgrade=2"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart.new_bootstrap=True"
"rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade"
"rptest.tests.cluster_features_test.FeaturesNodeJoinTest.test_synthetic_old_node_join"
"rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=True"
"rptest.tests.workload_upgrade_runner_test.RedpandaUpgradeTest.test_workloads_through_releases.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.cluster_bootstrap_test.ClusterBootstrapUpgrade.test_change_bootstrap_configs_after_upgrade.empty_seed_starts_cluster=False"
"rptest.tests.cluster_config_test.ClusterConfigLegacyDefaultTest.test_legacy_default.wipe_cache=False"
"rptest.tests.cluster_features_test.FeaturesMultiNodeTest.test_license_upload_and_query"
"rptest.tests.cluster_features_test.FeaturesUpgradeAssertionTest.test_upgrade_assertion"
"rptest.tests.pandaproxy_test.BasicAuthUpgradeTest.test_upgrade_and_enable_basic_auth.base_release=.22.3.next_release=.23.1"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_invalid_destination.num_to_upgrade=2"
"rptest.tests.topic_creation_test.CreateTopicUpgradeTest.test_retention_config_on_upgrade_from_v22_2_to_v22_3.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.transactions_test.GATransaction_v22_1_UpgradeTest.upgrade_coordinator_test"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40235#018b8b12-6e18-4b90-8a98-df658a37a2f7: "rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off"
"rptest.tests.cluster_bootstrap_test.ClusterBootstrapUpgrade.test_change_bootstrap_configs_after_upgrade.empty_seed_starts_cluster=True"
"rptest.tests.cluster_features_test.FeaturesNodeJoinTest.test_synthetic_too_new_node_join"
"rptest.tests.cluster_config_test.ClusterConfigLegacyDefaultTest.test_legacy_default.wipe_cache=True"
"rptest.tests.cluster_features_test.FeaturesMultiNodeUpgradeTest.test_rollback"
"rptest.tests.controller_snapshot_test.ControllerSnapshotTest.test_upgrade_compat"
"rptest.tests.controller_snapshot_test.ControllerSnapshotPolicyTest.test_upgrade_auto_enable"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_and_upgrade"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_finishes_after_manual_cancellation.delete_topic=True"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=5000"
"rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_flipping_decommission_recommission.node_is_alive=True"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_deletion_stops_move.num_to_upgrade=2"
"rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade_with_rollback.upgrade_after_rollback=False"
"rptest.tests.upgrade_test.UpgradeFromPriorFeatureVersionCloudStorageTest.test_rolling_upgrade.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.partition_movement_test.PartitionMovementTest.test_moving_not_fully_initialized_partition.num_to_upgrade=2"
"rptest.tests.topic_creation_test.CreateTopicUpgradeTest.test_retention_upgrade_with_cluster_remote_write.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.transactions_test.GATransaction_v22_1_UpgradeTest.upgrade_topic_test"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40311#018b905d-db3b-499d-b8c7-f9884842adf2: "rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type=CloudStorageType.S3.test_case=.TS_Read==True.AdjacentSegmentMergerReupload==True"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40311#018b905d-db42-4df8-8ced-824474b78247: "rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type=CloudStorageType.ABS.test_case=.TS_Read==True.AdjacentSegmentMergerReupload==True"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40311#018b905d-db45-4711-929c-5ad180e97c75: "rptest.tests.offset_for_leader_epoch_test.OffsetForLeaderEpochTest.test_offset_for_leader_epoch_transfer"

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/40311#018b905d-db3e-4d6b-be5b-61bb7c0b881e: "rptest.tests.rpk_registry_test.RpkRegistryTest.test_produce_consume_proto"

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/42192#018c3450-60bd-4327-b78f-6c1cdfb0b985:
```
"rptest.tests.recovery_mode_test.RecoveryModeTest.test_recovery_mode"
```
<!--5b8361be-5017-4db0-ace7-438cf2ddf110-->

andrwng commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,574 @@
+    }
+}
+
+ss::future<result<segment_upload::subscan_res>> segment_upload::subscan(
```
nit: gate and logger aside, feels like this should be a static storage utility method

andrwng reviewed (commented):
It seems like we're at most doing a subscan of two segments. Is it possible that the upload range covers multiple segments? If so, why don't we need to accumulate the contents of the intermediate segments?

Maybe I'm missing something, but it seems possible that there are no data records in the left or right subscan, but there are entire segments in the middle that have offsets to upload and data timestamps that get skipped.

Curious whether you considered accumulating over the entire upload range rather than just left and right, and found it to be insufficient. Even if it'd be more IO, it feels like it'd be worth avoiding the complexity and introducing new file-based dependencies (figuring out file size, etc) on the storage layer

dotnwat commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,774 @@
+            auto dli = dynamic_cast<const storage::disk_log_impl*>(log.get());
+            auto base_it = dli->segments().lower_bound(range.base);
+            std::vector<ss::lw_shared_ptr<storage::segment>> segments;
+            for (auto it = base_it; it != dli->segments().end(); it++) {
```
I was under the impression that we are removing the `log::segments()` abstraction.

Lazin commented on src/v/archival/async_data_uploader.h:
```diff
@@ -0,0 +1,244 @@
+    /// Scan partition to find target offset.
+    /// Returned 'subscan_res' struct contains size of the
+    /// scanned region on disk, number of batches/records,
+    /// the timestamp and offset of the 'target'.
+    /// The timestamp is a data timestamp (if the target is a config batch
+    /// the subscan will scan until the data batch is found and use first data
```
updated the comment

Lazin commented on src/v/archival/async_data_uploader.h:
```diff
@@ -0,0 +1,244 @@
+    ss::future<result<subscan_res>> subscan(
+      ss::lw_shared_ptr<cluster::partition> part,
+      inclusive_offset_range range,
+      size_t initial_offset,
```
fixed

Lazin commented on src/v/archival/async_data_uploader.h:
```diff
@@ -0,0 +1,244 @@
+      bool target_is_last_offset,
+      model::timeout_clock::time_point deadline) noexcept;
+
+    /// Calculate upload size using segment indexes
+    ss::future<result<cloud_upload_parameters>> compute_upload_parameters(
+      inclusive_offset_range range, model::timeout_clock::time_point deadline);
```
Agree, but at the moment this is not possible.

Lazin commented on src/v/archival/async_data_uploader.h:
```diff
@@ -0,0 +1,244 @@
+    ss::future<result<cloud_upload_parameters>> compute_upload_parameters(
+      inclusive_offset_range range, model::timeout_clock::time_point deadline);
+
+    // Scan te beginning of the segment range
+    ss::future<result<subscan_res>> subscan_left(
+      std::vector<ss::lw_shared_ptr<storage::segment>> segments,
+      model::offset range_base,
+      model::timeout_clock::time_point deadline);
+
+    /// Scan the end of the segment range
+    ss::future<result<subscan_res>> subscan_right(
+      std::vector<ss::lw_shared_ptr<storage::segment>> segments,
+      model::offset range_last,
+      model::timeout_clock::time_point deadline);
```
We have to upload an offset range that may span multiple segments and it may also start and stop in the middle of the segment. We have to find an upload candidate based on start offset and end offset of the upload or based on start offset and desired size of the upload. The algorithm is

- call `subscan_left` to find the exact beginning of the upload.
- call `subscan_right` to find the end of the upload.
- compute size and metadata of the upload based on two subscan results and segments in the middle.

The subscan uses segment index to find index entry closest to the target and then it scans up to the target to calculate precise file location. It may also scan past the target to find the data timestamp.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/43486#018cdab7-63a6-4ab1-a8c4-e6850fa3fbcc
<!--762e07b2-e6cc-4638-98b5-03dbd03b822b-->

Lazin commented:
I updated the PR. Removed all scanning from the `segment_upload` class and added `offset_range_size` method to the `disk_log_impl`. The `offset_range_size` can be used to find the size of the offset range on disk (or when its uploaded to S3 without reading all data first. It works similarly to previous `segment_upload` implementation. It uses index to find beginning and the end of the offset range and scans a little. It also has two overloads. One uses inclusive offset range. Another one uses offset and size limits.

Lazin commented:
The timestamp handling was removed. Previously, the `segment_upload` was able to tell what is the first/last data timestamp in the uploaded region. This was removed. Instead of doing this we will find the timestamps during the upload process (we have to look at every record batch during the upload anyway). This eliminates some corner cases (e.g. if the uploaded segment doesn't have any data batches we can just use timestamps from the previous segment in the manifest).

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1733,446 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
       .then([this](auto rdr) { return _readers_cache->put(std::move(rdr)); });
 }
 
+namespace details {
```
Thank you for pulling this logic into the storage layer! Now there are a couple of clearly separate pieces to this PR: local storage changes, and archival changes. Would you mind splitting them into separate PRs, where the latter depends on the former? That will make them both more reviewer-friendly

Lazin replied:
done

andrwng replied:
Thank you!

andrwng commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,374 @@
+    // Create a log reader config to scan the uploaded offset
+    // range. We should skip the batch cache.
+    storage::log_reader_config reader_cfg(
+      params.value().offsets.base,
+      params.value().offsets.last,
```
Should be using `_params`, given the move at L259?

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1733,446 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+
+                // We're including only part of the file so the total size
+                // of the range has to be adjusted
+                current_size
+                  -= (it->get()->file_size() - last_index_entry->filepos);
```
Maybe this is true with the current implementation, but doesn't this yield slightly inaccurate results? If the uploaded range of offsets isn't aligned to the granularity of the index, doesn't this fudge the resulting size a bit in a way that results in sizes that are slightly off?

Lazin replied:
Here we're stopping at the index boundary. The idea is that when we're doing normal uploads we have a last uploaded offset and the target size of the segment to upload (which is usually in megabytes). We will find the first offset of the upload precisely by using index and then scanning. But in order to find the last offset we don't need such precision. We can just use index to find some offset that justifies our requirements. If it's off by few hundred bytes it's not a problem. 

The good thing it that we don't need to scan the segment. This has the added benefit. After the first segment is uploaded all subsequent uploads start on an index boundary which means that we don't need to scan segments in the future. All subsequent uploads will begin and end on index boundaries.

andrwng commented on src/v/archival/async_data_uploader.cc:
```diff
@@ -0,0 +1,374 @@
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
```
nit: now that there is a public static initializer and no public constructor, WDYT about adding the stream as an argument to the `segment_upload` constructor. Perhaps we can run the bits in `initialize()` first to construct a stream, and then pass it in as a member to the `segment_upload`. That way there's never a time where `segment_upload` isn't ready for action, and we can remove the calls to `throw_if_not_initialized()`

andrwng reviewed (commented):
Thank you for decoupling the storage bits in this PR! It was very helpful in review.

I'm still a bit unsure about how accurate the size-finding pieces need to be. Once that's settled, it'd also be very much appreciated if the storage bits of this PR could be put in a separate PR.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1969,6 +1969,198 @@ ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
     };
 }
 
+ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
```
Can we please break stuff like this out into separate PRs along with unit tests? IIUC this type of change is a separate, pre-requisite for the upload mechanism work?

Lazin replied:
done

Lazin commented on src/v/model/timeout_clock.h:
```diff
@@ -35,4 +35,13 @@ time_from_now(model::timeout_clock::duration d) {
 }
 
+inline model::timeout_clock::duration
+duration_to(model::timeout_clock::time_point deadline) {
```
done

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1733,446 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+} // namespace details
+
+ss::future<size_t> disk_log_impl::get_file_offset(
+  ss::lw_shared_ptr<segment> s,
```
I guess it's to make the actual usage more intuitive, that this is specifically referring to the file position within the given segment, even if the segment isn't used in this context.

That said, maybe this method would be more intuitive if it took only two offsets (and I guess the inclusive bit), and the accounting/addition were left to the caller.

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+        co_await std::move(reader).consume(acc, model::no_timeout);
+    } catch (...) {
+        last_err = std::current_exception();
+        vlog(stlog.error, "Error detected while consuming {}", last_err);
```
why do you need `last_err variable` instead of just `throw;` here?

Lazin replied:
it made sense before I moved this code to the storage layer, good catch
fixed

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+      last,
+      offsets());
+    // build the collection
+    auto segments = [&] {
```
nit nit nit: const auto?

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+          segments.front()->offsets());
+        throw std::invalid_argument("First offset is out of range");
+    }
+    if (segments.back()->offsets().committed_offset < last) {
```
doesn't this need to be `<` or `<=` depending on boundary_type?

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+    std::vector<ss::future<ss::rwlock::holder>> f_locks;
+    f_locks.reserve(segments.size());
+    for (auto& s : segments) {
+        f_locks.emplace_back(s->read_lock());
```
any chance of deadlocks here or we assume locks are always/everwhere taken in the increasing segment order?

dotnwat replied:
> yep

which one of @nvartolomei questions are you responding to?

nvartolomei commented on src/v/storage/disk_log_impl.h:
```diff
@@ -172,6 +195,14 @@ class disk_log_impl final : public log {
     friend class disk_log_builder;  // for tests
     friend std::ostream& operator<<(std::ostream& o, const disk_log_impl& d);
 
+    /// Compute file offset of the batch inside the segment
```
From what I'm reading index_entry is just a hint? Shall we make it an optional rather than using {file_pos=0,offset=segment.base_offset} when we don't have it? get_file_offset already has the segment so it can compute base offest itself (if needed) when index_entry is not present


```
    } else {
        // Scan from the beginning of the segment.
        left_entry.filepos = 0;
        left_entry.offset = segments.front()->offsets().base_offset;
        vlog(
          stlog.debug,
          "Scanning (left) log segment {} from the start",
          segments.front()->offsets());
    }
    auto left_scan_bytes = co_await get_file_offset(
      segments.front(),
      left_entry,
      first,
      model::boundary_type::exclusive,
      io_priority);
```

nvartolomei replied:
It would also simplify the caller. You can get rid completely of the left_entry variable

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+        current_size = 0;
+    }
+
+    if (segments.empty()) {
```
why we are keeping this collection around until the end of the method?

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+        current_size = 0;
+    }
+
+    co_return offset_range_size_result_t{
```
i believe this should either be an optional or lines like `current_size=0` be removed and let the caller decide if the result is too small and should be thrown away

let's have one responsible site for this. less code. less brain energy needed to understand who and why does things

nvartolomei replied:
You still left `target.min_size`. I was suggesting that this could be removed completely and the above ```
    if (current_size < target.min_size) {
        vlog(stlog.debug, "Discovered offset range is not large enough");
        last_included_offset = model::offset{};
        current_size = 0;
    }
```

with it letting the caller do the if statement.

Lazin replied:
changed return type to optional

nvartolomei reviewed (commented):
Few commends. Haven't fully reviewed `offset_range_size` override with size requirements. Will come back to it.

nvartolomei commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -3542,3 +3545,918 @@ FIXTURE_TEST(test_skipping_compaction_below_start_offset, log_builder_fixture) {
+};
+
+FIXTURE_TEST(test_offset_range_size, storage_test_fixture) {
+#ifdef NDEBUG
```
why not run this in debug too? maybe with less "num_test_cases" if runtime is too prohibitive

nvartolomei replied:
i'd expect more invariants being asserted in underlying code/iterators/etc in debug mode

Lazin replied:
done

nvartolomei commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -3542,3 +3545,918 @@ FIXTURE_TEST(test_skipping_compaction_below_start_offset, log_builder_fixture) {
+};
+
+FIXTURE_TEST(test_offset_range_size2, storage_test_fixture) {
+#ifdef NDEBUG
```
ditto

nvartolomei commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -3542,3 +3545,918 @@ FIXTURE_TEST(test_skipping_compaction_below_start_offset, log_builder_fixture) {
+};
+
+FIXTURE_TEST(test_offset_range_size_compacted, storage_test_fixture) {
+#ifdef NDEBUG
```
ditto

nvartolomei commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -3542,3 +3545,918 @@ FIXTURE_TEST(test_skipping_compaction_below_start_offset, log_builder_fixture) {
+
+FIXTURE_TEST(test_offset_range_size_incremental, storage_test_fixture) {
+#ifdef NDEBUG
+    // This test attempts to consume the log incrementally using overload
```
This is great cover letter for the test. Since these tests are not the most obvious, I and future readers will greatly appreciate a cover letter on all of them.

Lazin replied:
done

Lazin commented:
Test failures is https://github.com/redpanda-data/redpanda/issues/16308

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,474 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+    if (
+      first_segment_offsets.base_offset <= first
+      && first_segment_offsets.committed_offset >= first) {
+        // The segment is first partially We're doing subscan here but
```
nit: partial sentence?

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,474 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+        // only the first upload will have to do scanning to find the
+        // offset.
+        auto ix_res = first_segment->index().find_nearest(first);
+        //
```
nit: missing comment?

andrwng commented on src/v/storage/segment_index.h:
```diff
@@ -153,6 +153,12 @@ class segment_index {
+    std::optional<entry> find_above_size_bytes(size_t distance);
+    /// Find entry by file offset (the value will undershoot or find precise
+    /// match)
+    std::optional<entry> find_below_size_bytes(size_t distance);
```
Bumping this question

Lazin replied:
it's not used at the moment

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,474 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+    auto first_segment = *base_it;
+    size_t first_segment_size = first_segment->file_size();
+    auto first_segment_offsets = first_segment->offsets();
+    auto r_lock = co_await first_segment->read_lock();
```
In case it's unintentional, this lock isn't dropped until the end of this method

Lazin commented on src/v/storage/segment_index.h:
```diff
@@ -153,6 +153,12 @@ class segment_index {
+    /// Find entry by file offset (the value may overshoot or find precise
+    /// match)
+    std::optional<entry> find_above_size_bytes(size_t distance);
+    /// Find entry by file offset (the value will undershoot or find precise
+    /// match)
```
fixed that

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,474 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+
+    auto holders = co_await ss::when_all_succeed(
+      std::begin(f_locks), std::end(f_locks));
+
```
It does take the write lock, but it drops it once the segment is closed, after which even if we have the read lock, I don't think we'll be able to read the segment

Lazin commented:
CI failure is unrelated https://github.com/redpanda-data/redpanda/issues/16402

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1735,520 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+              current_size);
+
+            size_t truncate_after = 0;
+            if (num_segments == 1) {
+                // There are two cases here.
+                // 1. The segment has at least target_size bytes after
+                // base_file_pos,
+                // 2. The segment is too small. In this case we need to clamp
+                // the result.
+                truncate_after = first_segment_file_pos + target.target_size;
+                truncate_after = std::clamp(
+                  truncate_after,
+                  first_segment_file_pos,
+                  it->get()->file_size());
+            } else {
+                // In this case we need to find the truncation point
+                // always starting from the beginning of the segment.
+                //
+                // prev is guaranteed to be smaller than target_size
+                // because we reached this branch.
+                auto prev = current_size - it->get()->file_size();
+                auto delta = target.target_size - prev;
+                truncate_after = delta;
+            }
```
I'm a little confused about why this has to be two distinct cases. In either case, isn't the idea to remove `current_size - target.target_size` bytes? And aren't we guaranteed that it's this segment that pushed us over the top?

Would something like this work:

```
auto cur_seg_pos = it->get()->file_size(); // Above calculation `current_size` has assumed that we've accepted to the end of the file, and now we need to truncate.
auto truncate_by = current_size - target.target_size;
auto truncate_after = cur_seg_pos - truncate_by;
```

andrwng replied:
>we need to use the size of the truncated prefix to locate the end of the offset range.

Right, but doesn't `current_size` already account for the missing (or not) prefix?

I suspect these are the same, translating my suggestion, ignoring the clamp:

case of 1 segment:
cur_seg_pos = first_seg_file_size
current_size = first_seg_file_size - first_seg_file_pos
truncate_by = first_seg_file_size - first_seg_file_pos - target.target_size
truncate_after = first_seg_file_size - (first_seg_file_size - first_seg_file_pos - target.target_size)
... = first_seg_file_pos + target.target_size (same as L2147, and sure we can clamp after)

case of more than 1 segment:
cur_seg_pos = iter->file_size()
current_size = size of previous segments + iter->file_size()
truncate_by = size of previous segments + iter->file_size() - target.target_size
truncate_after = iter->file_size() - (size of previous segments + iter->file_size() - target.target_size)
... = target.target_size - size of previous segments (same as L2160)

After writing it all out, I'm more convinced these are the same, though it's somewhat a nit. I do think it makes the code more readable though.

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1735,520 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+                last_included_offset = it->get()->offsets().committed_offset;
+            }
+            break;
+        } else if (current_size > target.min_size) {
+            vlog(
+              stlog.debug,
+              "Setting offset range to {} - {}",
+              first,
+              it->get()->offsets().committed_offset);
+            // We can include full segment to the list of segments
+            last_included_offset = it->get()->offsets().committed_offset;
+            continue;
```
Why is this gated on `current_size > target.min_size`? Doesn't the `current_size < target.min_size` check below kind of encompass that? Or put another way, if the first segments we iterated through didn't meet the `min_size` requirement, why is it important to avoid setting this?

andrwng replied:
Got it, thanks for the explanation. IMO it makes it more confusing leaving it in, especially because there are so many branches in this code already

andrwng reviewed (commented):
Sorry the reviews here have dragged on here. Structurally, and ignoring side conversations about the right interface between storage and archival, I think this is looking pretty good.

While I appreciate that this has already split up the storage/archival changes, it'd really help as a reviewer to break things down even further. One approach to consider is to split out PRs into independtly testable units. For instance, this PR could probably have been split into:
- new method for the segment_index
- batch cache option for the log reader
- get_file_offset() + maybe batch_size_accumulator
- offset_range_size(start, last)
- offset_range_size(start, size)

It's much harder to introduce bugs when the PRs are small. That said, given there's already quite a lot of review on this PR, I can understand if you would prefer to punt on this.

andrwng reviewed (approved):
Remaining comments are pretty much nits. I know it's been a long process, but thanks for continuing to improve this code!

dotnwat commented on src/v/storage/segment_index.h:
```diff
@@ -153,6 +153,12 @@ class segment_index {
     std::optional<entry> find_nearest(model::timestamp);
+    /// Find entry by file offset (the value may overshoot or find precise
+    /// match)
+    std::optional<entry> find_above_size_bytes(size_t distance);
```
can this be `const`?

dotnwat commented on src/v/storage/segment_index.h:
```diff
@@ -153,6 +153,12 @@ class segment_index {
+    std::optional<entry> find_above_size_bytes(size_t distance);
+    /// Find entry by file offset (the value will undershoot or find precise
+    /// match)
+    std::optional<entry> find_below_size_bytes(size_t distance);
```
can this be const?

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,244 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+struct batch_size_accumulator {
+    ss::future<ss::stop_iteration> operator()(model::record_batch b) {
+        vassert(
+          result_size_bytes != nullptr,
```
it seems to me like you could add a constructor to this struct and not pay this cost on each operator() invocation?

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1731,6 +1734,244 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+    auto reader = co_await make_reader(reader_cfg);
+
+    try {
+        co_await std::move(reader).consume(acc, model::no_timeout);
```
i thought that consume() allowed the consumer to return a value, which would mean, you probably don't need to pass in a pointer to size_bytes in order to get size_bytes back out after the consumer is done processing data from the reader?

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1972,6 +1972,203 @@ ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
+  offset_range_size_requirements_t target,
+  ss::io_priority_class io_priority) {
+    if (
+      false
```
this looks like it's in the PR unintentionally?

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1972,6 +1972,203 @@ ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
+    for (auto it = base_it; it < _segs.end(); it++) {
+        const auto& offsets = it->get()->offsets();
+        auto r_lock = co_await it->get()->read_lock();
+        if (offsets.base_offset <= first && offsets.committed_offset >= first) {
```
generally speaking anytime we have a reference to something accessed via an iterator held across `co_await` calls we should see a comment that explains why it is safe. can you add that here?

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1972,6 +1972,203 @@ ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
+            }
+            auto file_pos = co_await get_file_offset(
+              *it, index_entry, first, false, io_priority);
+            auto sz = it->get()->file_size() - file_pos;
```
how is any of this safe w.r.t. iterator invalidations?

dotnwat commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -3667,3 +3668,183 @@ FIXTURE_TEST(test_offset_range_size, storage_test_fixture) {
 };
+
+FIXTURE_TEST(test_offset_range_size2, storage_test_fixture) {
+#ifdef NDEBUG
```
there should definitely be a comment explaining why this doesn't work in debug mode.

dotnwat commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -3848,3 +3848,175 @@ FIXTURE_TEST(test_offset_range_size2, storage_test_fixture) {
 };
+
+FIXTURE_TEST(test_offset_range_size_compacted, storage_test_fixture) {
+#ifdef NDEBUG
```
there should definitely be a comment explaining why this doesn't work in debug mode.

dotnwat commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -4020,3 +4031,262 @@ FIXTURE_TEST(test_offset_range_size_compacted, storage_test_fixture) {
 };
+
+FIXTURE_TEST(test_offset_range_size2_compacted, storage_test_fixture) {
+#ifdef NDEBUG
```
why doesn't this work in debug mode
